distributed/comm/ws.py

```diff
@@ -0,0 +1,372 @@
+                **self.handshake_options,
+            },
+        )
+        n = struct.pack("Q", len(frames))
+        try:
+            await self.sock.write_message(n, binary=True)
+            for frame in frames:
+                await self.sock.write_message(ensure_bytes(frame), binary=True)
+        except WebSocketClosedError as e:
+            raise CommClosedError(e)
+
+        return sum(map(nbytes, frames))
+
+    async def close(self):
+        if not self.sock.close_code:
+            self._finalizer.detach()
+            self.sock.close()
+        self._closed = True
+
+    def abort(self):
+        if not self.sock.close_code:
+            self._finalizer.detach()
+            self.sock.close()
+        self._closed = True
+
+    def closed(self):
+        return not self.sock or self.sock.close_code or self._closed
+
+    @property
+    def local_address(self):
+        return f"{self.prefix}{self.sock.parsed.netloc}"
+
+    @property
+    def peer_address(self):
+        return f"{self.prefix}{self.sock.parsed.netloc}"
+
+    def _read_extra(self):
+        pass
+
+    @property
+    def extra_info(self):
+        return self._extra
+
+
+class WSS(WS):
+    prefix = "wss://"
+
+    def _read_extra(self):
+        WS._read_extra(self)
+        sock = self.sock.stream.socket
+        if sock is not None:
+            self._extra.update(peercert=sock.getpeercert(), cipher=sock.cipher())
+            cipher, proto, bits = self._extra["cipher"]
+            logger.debug(
+                "TLS connection with %r: protocol=%s, cipher=%s, bits=%d",
+                self._peer_addr,
+                proto,
+                cipher,
+                bits,
+            )
+
+
+class WSListener(Listener):
+    prefix = "ws://"
+
+    def __init__(
+        self,
+        address: str,
+        handler: Callable,
+        deserialize: bool = False,
+        **connection_args,
+    ):
+        if not address.startswith(self.prefix):
+            address = f"{self.prefix}{address}"
+
+        self.ip, self.port = parse_host_port(address, default_port=0)
+        self.handler = handler
+        self.deserialize = deserialize
+        self.connection_args = connection_args
+        self.bound_address = None
+        self.server_args = self._get_server_args(**connection_args)
+
+    def _get_server_args(self, **connection_args):
+        return {}
+
+    @property
+    def address(self) -> str:
+        return f"{self.prefix}{self.ip}:{self.port}"
+
+    async def start(self):
+        routes = [
+            (
+                r"/",
+                WSHandler,
+                {
+                    "handler": self.handler,
+                    "deserialize": self.deserialize,
+                    "listener": weakref.ref(self),
+                },
+            )
+        ]
+        try:
+            self.server = self.handler.__self__.http_server
+            self.handler.__self__.http_application.add_handlers(r".*", routes)
```

Comment:
I'm trying to reuse the dashboard's server, but in the context of a secure websocket connection I ran into an issue trying to do something like: 
```python
In [1]: from dask.distributed import Client
In [2]: client = Client(protocol="wss", security=True)
```
Looks like the dashboard ssl context (if present) is coming from `dask.config.get("distributed.scheduler.dashboard.tls.key") ...` and not from the self-signed cert. Since those were different servers this wasn't an issue, but now that I'm trying to use a single one for both dashboard and comms this is a problem.
The `start_http_server` method has the `ssl_options` kwarg (https://github.com/dask/distributed/blob/master/distributed/node.py#L99-L121) but it's not used by the scheduler/workers. And even if they did, those might get overridden from the config file.
Playing around with the scheduler code I got it to work with the self-signed cert by changing the `start_http_server` call from
```python
self.start_http_server(routes, dashboard_address, default_port=8787)
```
to
```python
self.start_http_server(
    routes, dashboard_address, default_port=8787, 
    ssl_options=self.connection_args["ssl_context"]
)
```
But in a more production settings this will probably not work.

I'm struggling to find a good approach with minimal changes.

Thoughts?

Reply:
@jcrist if you have time do you have thoughts  on this?

Reply:
@jacobtomlinson you also seem like someone who might be able to help here.

Reply:
Ping @jacobtomlinson or @quasiben in case either of you have time/expertise to share here.

Reply:
Am I right in thinking that it is possible to configure different sets of certs for the dask comm and the dashboard web server?

And if so has anyone seen folks intentionally use different certs for the comm and dashboard in the past? If not perhaps some work so consolidate this would make things easier here.

Reply:
> Am I right in thinking that it is possible to configure different sets of certs for the dask comm and the dashboard web server?

You are absolutely right, which is the problem I'm running into trying to share the same server.
> And if so has anyone seen folks intentionally use different certs for the comm and dashboard in the past?

I can't really speak to this. @mrocklin ?

> If not perhaps some work so consolidate this would make things easier here.

Definitely :)

Reply:
> And if so has anyone seen folks intentionally use different certs for the comm and dashboard in the past?

Well, no one has used this yet, so no.  However, maybe a scenario here would be that we want the Dask communication to be secure, but don't mind if the dashboard is publicly accessible.

Reply:
Most users securing dask with TLS are using self-signed-certs that won't be usable via the browser (the browser won't have a compatible CA). As such, it's important to have the option for using different certs for Dask communication and the dashboard, so the browser can still work (either no TLS, or a cert the browser will accept).

Reply:
Maybe a dumb question, but is it possible to have one HTTP server that uses different certs for different routes?

Reply:
The routes are encrypted, so by the time they get to the server the certs have to already be chosen.

Perhaps a good design here would be:

- By default spin up a 2nd server when using Websocket comms, unless the dask comm port is specified as the same as the dashboard port (then share the server).
- If sharing the server, use the dask comm certs first (if configured), fall back to the dashboard certs (if configured), then no certs.

Reply:
@marcosmoyano , should you find yourself with free time before you head out (low chance, I know) this seems like a sensible approach (actually, I have no idea, but if Jim and Jacob both agree then I have high confidence).

Reply:
Ok, so if the ports don't match behave like any other comm and fire up a 2nd server. +1
When the ports match try to share the server. It looks like the dashboard server fires up before the scheduler comm so I would have bring some protocol awareness before that happens to do the protocol/port checking and make sure that the `ssl_options` argument is respected. I'm not sure how clean it'll end up. I'll play with it and see how that goes. A couple of high level questions 
* What happens if the comm port is the same as the dashboard but no protocol was specified? Should it raise an exception? Default to using websockets?
* Looks like the ssl context for the dashboard server is created with `purpose=ssl.Purpose.SERVER_AUTH` while all other comms start with `purpose=ssl.Purpose.CLIENT_AUTH` (the naming of ssl.Purpose is confusing). If sharing the server those would have to match. Also, if anyone can explain why the dashboard server start with `SERVER_AUTH` that would be great. 
* When sharing the server, the dashboard won't work with self-signed certs. Should it default to no certs at all? Does the dashboard take precedence over having security communication within the cluster? Based on @jcrist comment it looks like it. Just verifying.

Alternatively, this could be handled through documentation. "If using the secure websocket protocol, Dask shares a single server for the dashboard and the comm so both certs have to match" 🤷🏼‍♂️

Reply:
> Most users securing dask with TLS are using self-signed-certs that won't be usable via the browser
>> If sharing the server, use the dask comm certs first (if configured), fall back to the dashboard certs (if configured), then no certs.

So self-signed-certs from comms will take precedence over the dashboard. I think the dashboard will still work, but won't be over ssl.

Security concern: dask comms use bidirectional authentication with `ssl_context.verify_mode = ssl.CERT_REQUIRED` while the dashboard uses `ssl_context.verify_mode = ssl.CERT_NONE`. For the dashboard to keep working we need `ssl.CERT_NONE` but when sharing the server that means that the scheduler comm will also have `ssl.CERT_NONE`. Thoughts?